openssl/src/evp.rs

```diff
@@ -0,0 +1,281 @@
+//!     let mut enc_len = seal.update(secret, &mut encrypted).unwrap();
+//!     enc_len += seal.finalize(&mut encrypted[enc_len..]).unwrap();
+//! }
+//! ```
+use error::ErrorStack;
+use ffi;
+use foreign_types::ForeignType;
+use libc::{c_int, c_uchar};
+use pkey::{PKey, Private, Public};
+use std::cmp;
+use symm::Cipher;
+use {cvt, cvt_p};
+
+/// Represents a EVP_Seal context.
+pub struct EvpSeal {
+    ctx: *mut ffi::EVP_CIPHER_CTX,
+    block_size: usize,
+    iv: Vec<u8>,
+    ek: Vec<Vec<u8>>,
+}
+
+/// Represents a EVP_Open context.
+pub struct EvpOpen {
+    ctx: *mut ffi::EVP_CIPHER_CTX,
+    block_size: usize,
+}
+
+impl EvpSeal {
+    /// Creates a new `EvpSeal`.
+    pub fn new(t: Cipher, pub_keys: Vec<PKey<Public>>) -> Result<EvpSeal, ErrorStack> {
+        unsafe {
+            let ctx = cvt_p(ffi::EVP_CIPHER_CTX_new())?;
+            let mut ek = Vec::new();
+            let mut pubk: Vec<*mut ffi::EVP_PKEY> = Vec::new();
+            let mut my_ek = Vec::new();
+            for key in &pub_keys {
+                let mut key_buffer: Vec<c_uchar>;
+                key_buffer = vec![0; ffi::EVP_PKEY_size(key.as_ptr()) as usize];
+                let tmp = key_buffer.as_mut_ptr();
+                my_ek.push(key_buffer);
+                ek.push(tmp);
+                pubk.push(key.as_ptr());
+            }
+            let mut iv_buffer: Vec<c_uchar> =
+                vec![0; ffi::EVP_CIPHER_iv_length(t.as_ptr()) as usize];
+            let mut ekl: Vec<c_int> = vec![0; ek.len()];
+
+            cvt(ffi::EVP_SealInit(
+                ctx,
+                t.as_ptr(),
+                ek.as_mut_ptr(),
+                ekl.as_mut_ptr(),
+                iv_buffer.as_mut_ptr(),
+                pubk.as_mut_ptr(),
+                pubk.len() as i32,
+            ))?;
+            Ok(EvpSeal {
+                ctx,
+                block_size: t.block_size(),
+                iv: iv_buffer,
+                ek: my_ek,
+            })
+        }
+    }
+
+    /// Return used initialization vector.
+    pub fn iv(&self) -> &Vec<u8> {
+        &self.iv
+    }
+
+    /// Return vector of keys encrypted by public key.
+    pub fn ek(&self) -> &Vec<Vec<u8>> {
+        &self.ek
+    }
+
+    /// Feeds data from `input` through the cipher, writing encrypted bytes into `output`.
+    ///
+    /// The number of bytes written to `output` is returned. Note that this may
+    /// not be equal to the length of `input`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `output.len() < input.len() + block_size` where
+    /// `block_size` is the block size of the cipher (see `Cipher::block_size`),
+    /// or if `output.len() > c_int::max_value()`.
+    pub fn update(&mut self, input: &[u8], output: &mut [u8]) -> Result<usize, ErrorStack> {
+        unsafe {
+            let mut outl = output.len() as c_int;
+            let inl = input.len() as c_int;
+            cvt(ffi::EVP_EncryptUpdate(
```

Comment:
Shouldn't this use EVP_SealUpdate?

Reply:
In OpenSSL source code EVP_SealUpdate is macro, translating this function directly to EVP_EncryptUpdate. So as far I know I can't use it. [OpenSSL documentation](https://www.openssl.org/docs/man1.1.1/man3/EVP_SealInit.html) also states that EVP_SealUpdate have exactly the same properties as the EVP_EncryptUpdate.

Reply:
Ah ok cool.